File: apps/server/routers/plugins.py
from fastapi import APIRouter, HTTPException
from typing import List, Dict, Any
import os
import json
from pathlib import Path

router = APIRouter(prefix="/api/plugins", tags=["plugins"])

# Base path for plugins
PLUGINS_DIR = Path(__file__).parent.parent.parent.parent / "plugins"


def load_plugin_manifest(plugin_dir: Path) -> Dict[str, Any]:
    """Load a plugin manifest from its directory."""
    manifest_path = plugin_dir / "manifest.json"
    if not manifest_path.exists():
        return None

    with open(manifest_path, "r", encoding="utf-8") as f:
        return json.load(f)


def get_all_plugins() -> List[Dict[str, Any]]:
    """Get all available plugins."""
    plugins = []

    if not PLUGINS_DIR.exists():
        return plugins

    for item in PLUGINS_DIR.iterdir():
        if item.is_dir():
            manifest = load_plugin_manifest(item)
            if manifest:
                plugins.append(manifest)

    return plugins


@router.get("", response_model=List[Dict[str, Any]])
async def list_plugins():
    """List all available plugins."""
    return get_all_plugins()


@router.get("/{plugin_id}", response_model=Dict[str, Any])
async def get_plugin(plugin_id: str):
    """Get a specific plugin manifest."""
    plugin_dir = PLUGINS_DIR / plugin_id
    if not plugin_dir.exists():
        raise HTTPException(status_code=404, detail="Plugin not found")

    manifest = load_plugin_manifest(plugin_dir)
    if not manifest:
        raise HTTPException(status_code=404, detail="Plugin manifest not found")

    return manifest

File: apps/server-ts/src/routes/plugins.ts
/**
 * Plugin API Routes - Plugin manifest listing.
 *
 * Ported from Python apps/server/routers/plugins.py
 */

import { readdir } from "node:fs/promises";
import { join } from "node:path";
import { Hono } from "hono";
import { getPluginsDir, resolvePluginDir } from "../engine/plugin-loader";

const app = new Hono();

const PLUGINS_DIR = getPluginsDir();

async function loadPluginManifest(pluginDir: string): Promise<Record<string, unknown> | null> {
  const manifestPath = join(pluginDir, "manifest.json");
  try {
    const file = Bun.file(manifestPath);
    if (!(await file.exists())) return null;
    return await file.json();
  } catch {
    return null;
  }
}

let pluginCache: Record<string, unknown>[] | null = null;

async function getAllPlugins(): Promise<Record<string, unknown>[]> {
  if (pluginCache) return pluginCache;

  try {
    const entries = await readdir(PLUGINS_DIR, { withFileTypes: true });
    const results = await Promise.all(
      entries
        .filter((entry) => entry.isDirectory())
        .map((entry) => loadPluginManifest(join(PLUGINS_DIR, entry.name))),
    );
    pluginCache = results.filter(Boolean) as Record<string, unknown>[];
  } catch {
    pluginCache = [];
  }

  return pluginCache;
}

// List all plugins
app.get("/", async (c) => {
  return c.json(await getAllPlugins());
});

// Get specific plugin
app.get("/:pluginId", async (c) => {
  const pluginId = c.req.param("pluginId");
  const pluginDir = resolvePluginDir(pluginId);
  if (!pluginDir) {
    return c.json({ detail: "Invalid plugin ID" }, 400);
  }

  const manifest = await loadPluginManifest(pluginDir);
  if (!manifest) return c.json({ detail: "Plugin not found" }, 404);
  return c.json(manifest);
});

export { app as pluginRoutes };
